sandbox/cgroup/monitor.go

```
@@ -33,122 +33,190 @@ import (
 	channel chan<- string
 }
 
-var currentWatcher *inotifyWatcher = &inotifyWatcher{
-	wd:           nil,
-	pathList:     make(map[string]int32),
-	addWatchChan: make(chan *groupToWatch),
+var currentWatcher *inotifyWatcher = newInotifyWatcher(context.Background())
+
+func newInotifyWatcher(ctx context.Context) *inotifyWatcher {
+	ctx, cancel := context.WithCancel(ctx)
+	return &inotifyWatcher{
+		ctx:          ctx,
+		pathList:     make(map[string]uint),
+		addWatchChan: make(chan *groupToWatch),
+		doneChan:     make(chan struct{}),
+		done:         cancel,
+	}
 }
 
 func (gtw *groupToWatch) sendClosedNotification() {
-	defer func() {
-		if err := recover(); err != nil {
-			logger.Noticef("Failed to send Closed notification for %s", gtw.name)
-		}
+	go func() {
+		defer func() {
+			if err := recover(); err != nil {
+				logger.Noticef("cannot send closed notification for %s", gtw.name)
+			}
+		}()
+		gtw.channel <- gtw.name
 	}()
-	gtw.channel <- gtw.name
+}
+
+// Close stops the watcher and waits for it to finish.
+func (iw *inotifyWatcher) Close() {
+	iw.done()
+	<-iw.doneChan
+	for p := range iw.pathList {
+		iw.removePath(p)
+	}
+	iw.wd.Close()
 }
 
 func (iw *inotifyWatcher) addWatch(newWatch *groupToWatch) {
-	var folderList []string
-	for _, fullPath := range newWatch.folders {
+	for fullPath := range newWatch.folders {
 		// It's not possible to use inotify.InDeleteSelf in /sys/fs because it
 		// isn't triggered, so we must monitor the parent folder and use InDelete
-		basePath := path.Dir(fullPath)
+		basePath := filepath.Dir(fullPath)
 		if _, exists := iw.pathList[basePath]; !exists {
 			iw.pathList[basePath] = 0
 			if err := iw.wd.AddWatch(basePath, inotify.InDelete); err != nil {
-				target := &os.PathError{}
-				if !errors.As(err, &target) {
-					logger.Noticef("Error when calling AddWatch for path %s: %s", target.Path, target.Err)
-				}
+				// TODO propagate the error back to the caller
+				logger.Noticef("cannot add watch %v", err)
```

Even if we don't implement this now, I wonder what the mechanism ought to look like. Here we could just return the error and move the logging to the caller in the loop. The loop works in a separate goroutine, so one'd have to have some sort of asynchronous return path with another channel.

---

IMO simplest would be to send a `chan error` in the caller and expect a reply.